core/src/sumeragi/fault.rs

```diff
@@ -68,194 +66,55 @@ where
+    pub current_online_peers_by_public_key: Mutex<Vec<PublicKey>>,
+
+    pub incomming_message_sender: Mutex<std::sync::mpsc::Sender<Message>>,
+    pub incomming_message_receiver: Mutex<std::sync::mpsc::Receiver<Message>>,
 }
 
-impl<F: FaultInjection> SumeragiWithFault<F> {
-    fn from_configuration(
-        configuration: &SumeragiConfiguration,
-        events_sender: EventsSender,
-        wsv: WorldStateView,
-        transaction_validator: TransactionValidator,
-        telemetry_started: bool,
-        genesis_network: Option<GenesisNetwork>,
-        queue: Arc<Queue>,
-        broker: Broker,
-        kura: Arc<Kura>,
-        // network: Addr<IrohaNetwork>,
-    ) -> Result<Self> {
-        let network_topology = Topology::builder()
-            .at_block(EmptyChainHash::default().into())
-            .with_peers(configuration.trusted_peers.peers.clone())
-            .build()?;
-
-        Ok(Self {
-            key_pair: configuration.key_pair.clone(),
-            topology: network_topology,
-            peer_id: configuration.peer_id.clone(),
-            voting_block: None,
-            votes_for_blocks: BTreeMap::new(),
-            events_sender,
-            wsv: std::sync::Mutex::new(wsv),
-            txs_awaiting_receipts: HashMap::new(),
-            txs_awaiting_created_block: HashSet::new(),
-            votes_for_view_change: HashMap::new(),
-            commit_time: Duration::from_millis(configuration.commit_time_limit_ms),
-            tx_receipt_time: Duration::from_millis(configuration.tx_receipt_time_limit_ms),
-            block_time: Duration::from_millis(configuration.block_time_ms),
-            block_height: 0,
-            invalidated_blocks_hashes: Vec::new(),
-            telemetry_started,
-            transaction_limits: configuration.transaction_limits,
-            transaction_validator,
-            genesis_network,
-            queue,
-            broker,
-            kura,
-            // network,
-            actor_channel_capacity: configuration.actor_channel_capacity,
-            fault_injection: PhantomData,
-            gossip_batch_size: configuration.gossip_batch_size,
-            gossip_period: Duration::from_millis(configuration.gossip_period_ms),
-        })
-    }
+pub struct AtomicSumeragiState {
+    pub genesis_network: Option<GenesisNetwork>,
+    pub latest_block_hash: HashOf<VersionedCommittedBlock>,
+    pub latest_block_height: u64,
+    pub current_topology: Topology,
+}
 
-    /// Initializes sumeragi with the `latest_block_hash` and `block_height` after Kura loads the blocks.
-    pub fn init(&mut self, latest_block: HashOf<VersionedCommittedBlock>, block_height: u64) {
-        self.block_height = block_height;
-        self.topology.apply_block(latest_block);
+impl<F: FaultInjection> SumeragiWithFault<F> {
+    pub fn get_online_peer_keys(&self) -> Vec<PublicKey> {
+        self.current_online_peers_by_public_key
+            .lock()
```

Comment:
Why do we need this, and not just an `RwLock`, if all we're going to do is `clone` the `Vec`?

Reply:
Because holding locks is more complicated then not doing so. `current_online_peers_by_public_key` is not a big array so this is not a problem. Cloning all the time causes performance problems. Sharing data all the time also causes performance problems and *slows down development* because there is more to think about. My guiding principle is to avoid holding locks when I don't have to. Makes things easier.

Reply:
Yes. And allowing concurrent reading access makes programmers think it's okay to hold the lock. Because it avoids a clone and "that's good right?". In this case it is very important that the data is able to be updated. I'd rather avoid the bug of "sometimes the online peers won't update" that got introduced because someone thought they were being smart. You guys being so resistant to the mutex proves why it is good. It communicates very strongly. DO NOT HOLD THIS LOCK. DOING SO WILL BE BAD. A RWLock communicates a more seducive "Look at how smart I'm being by allowing multiple readers on the same data."

A codebase where one is very hesitant to hold locks leads to a more healthy view on multithreading. It's fast when threads don't communicate. When they have independent work to do. And the only reason for communicating between threads is to arrange for more work to be done in solitude.